File: evm/typescript/src/roles.ts
import type { ContractFunctionReturnType } from "viem";
import type { swigConfigAbi } from "./abi.js";
import { type Authority, decodeAuthority } from "./authorities.js";
import { uint } from "./codec.js";

export const ROOT_ROLE_ID = 0;
export type RoleData = ContractFunctionReturnType<
  typeof swigConfigAbi,
  "view",
  "getRole"
>;

export interface Role {
  readonly id: number;
  readonly authority: Authority;
  readonly actionCount: number;
}

/** Decode one getRole result without fetching actions or other wallet state. */
export function decodeRole(data: RoleData): Role {
  uint(data.id, 0xffffffff);
  uint(data.actionCount, 0xffff);
  return {
    id: data.id,
    authority: decodeAuthority(data.authority),
    actionCount: data.actionCount,
  };
}

File: evm/rust/src/roles.rs
use crate::{authorities::Authority, codec::CodecError, contracts::config::SwigConfig};

pub const ROOT_ROLE_ID: u32 = 0;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Role {
    pub id: u32,
    pub authority: Authority,
    pub action_count: u16,
}

impl Role {
    /// Decode one getRole result without fetching actions or other wallet state.
    pub fn decode(data: &SwigConfig::Role) -> Result<Self, CodecError> {
        Ok(Self {
            id: data.id,
            authority: Authority::decode(&data.authority)?,
            action_count: data.actionCount,
        })
    }
}
